### PR TITLE
fix: use LinkedHashSet (not HashSet) for deterministic CompositeDataGenerator 

### PR DESCRIPTION
**Pull request by TatuLund**
Fixes https://github.com/vaadin/flow/issues/9656

### CODE REVIEW

File: flow-data/src/main/java/com/vaadin/flow/data/provider/CompositeDataGenerator.java [hunk 1]
```diff
@@ -39,7 +39,7 @@
 public class CompositeDataGenerator<T>
         implements DataGenerator<T>, HasDataGenerators<T> {
 
-    private final Set<DataGenerator<T>> dataGenerators = new HashSet<>();
+    private final List<DataGenerator<T>> dataGenerators = new ArrayList<>();
```
Comment (joheriks) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/CompositeDataGenerator.java:42]: Does this change API slightly in that `addDataGenerator` is no longer idempotent? Would perhaps be safer to use `LinkedHashSet `, which has set semantics but preserves order.

Comment (TatuLund) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/CompositeDataGenerator.java]: True

**Comment by Artur-** [comment]
I am not saying tests are mandatory but if you leave them out, it means you are fine if somebody reverts the patch tomorrow. It should not be difficult to add a test that the order is preserved

**Comment by joheriks** [comment]
> Ok, did that, which required to make dataGenerators package protected.

It can be tested without the visibility modificiation, with just slightly blacker-box test. E.g. public method `destroyAllData` calls the `dataGenerators` in order. The test could register e.g. three `DataGenerator` instances which have side effect on a local mutable, e.g., appending `this` to a local list. Then the list can be asserted to be in order of registration.

File: flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java [hunk 2]
```diff
@@ -196,4 +196,14 @@ public void dataGeneratorRegistration_remove_dataIsDestroyed() {
                 CoreMatchers.not(CoreMatchers.hasItem("item1")));
     }
 
+    @Test
+    public void orderIsPreserved() {
+        CompositeDataGenerator<String> cdg = new CompositeDataGenerator<>();
+        DataGenerator<String> dg1 = (String, JsonObject) -> {};
+        DataGenerator<String> dg2 = (String, JsonObject) -> {};
+        List<DataGenerator<String>> expected = Arrays.asList(dg1,dg2);
```
Comment (joheriks) [inline flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java:204]: `Arrays` may need import.

File: flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java [hunk 3]
```diff
@@ -196,4 +196,14 @@ public void dataGeneratorRegistration_remove_dataIsDestroyed() {
...
+        cdg.addDataGenerator(dg1);
+        cdg.addDataGenerator(dg2);
+        assertEquals(expected,new ArrayList<>(cdg.dataGenerators));
```
Comment (joheriks) [inline flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java:207]: `Assert.assertEquals` or import.

**Review by joheriks — CHANGES_REQUESTED** [review]

**Comment by TatuLund** [comment]
> It can be tested without the visibility modificiation

The both approaches have their pros and cons, it is really hard to judge which is better.

**Review by joheriks — APPROVED** [review]

**Comment by joheriks** [comment]
> The both approaches have their pros and cons, it is really hard to judge which is better.

True, I'm fine with package-private. Just wanted to highlight the alternative.

File: flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java [hunk 4]
```diff
@@ -196,4 +200,14 @@ public void dataGeneratorRegistration_remove_dataIsDestroyed() {
                 CoreMatchers.not(CoreMatchers.hasItem("item1")));
     }
 
+    @Test
+    public void orderIsPreserved() {
```
Comment (mshabarov) [inline flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java:204]:
```suggestion
    public void addDataGenerator_orderIsPreserved() {
```

File: flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java [hunk 5]
```diff
@@ -196,4 +200,14 @@ public void dataGeneratorRegistration_remove_dataIsDestroyed() {
...
+        CompositeDataGenerator<String> cdg = new CompositeDataGenerator<>();
+        DataGenerator<String> dg1 = (String, JsonObject) -> {};
+        DataGenerator<String> dg2 = (String, JsonObject) -> {};
+        List<DataGenerator<String>> expected = Arrays.asList(dg1,dg2);
+        cdg.addDataGenerator(dg1);
+        cdg.addDataGenerator(dg2);
+        assertEquals(expected,new ArrayList<>(cdg.dataGenerators));
```
Comment (mshabarov) [inline flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java:211]:
```suggestion
        assertEquals(expected, new ArrayList<>(cdg.dataGenerators));
```

File: flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java [hunk 6]
```diff
@@ -196,4 +200,14 @@ public void dataGeneratorRegistration_remove_dataIsDestroyed() {
                 CoreMatchers.not(CoreMatchers.hasItem("item1")));
     }
 
+    @Test
+    public void orderIsPreserved() {
+        CompositeDataGenerator<String> cdg = new CompositeDataGenerator<>();
+        DataGenerator<String> dg1 = (String, JsonObject) -> {};
+        DataGenerator<String> dg2 = (String, JsonObject) -> {};
+        List<DataGenerator<String>> expected = Arrays.asList(dg1,dg2);
```
Comment (mshabarov) [inline flow-data/src/test/java/com/vaadin/flow/data/provider/CompositeDataGeneratorTest.java:208]:
```suggestion
        List<DataGenerator<String>> expected = Arrays.asList(dg1, dg2);
```

**Review by mshabarov — APPROVED** [review]
Small nits

**Review by mshabarov — CHANGES_REQUESTED** [review]
> > It can be tested without the visibility modificiation

> The both approaches have their pros and cons, it is really hard to judge which is better.

I would go with Johannes's approach (use `callOrder`) and keep the modifier `private`, because the package consists of a lot of classes with heavy data handling logic, and the risk of accidentally modify the internal state is high.

**Review by mshabarov — APPROVED** [review]

**Comment by mshabarov** [comment]
Cherry picked to `2.4`

**Pull request merged**